Build failure looks related
```
======================================================================
FAIL: corehq.apps.app_manager.tests.test_views:TestViews.test_get_apps_modules
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/vendor/lib/python2.7/site-packages/mock/mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/tests/test_views.py", line 244, in test_get_apps_modules
    self.assertEqual(len(apps_modules), 2, 'get_apps_modules should only return normal Applications')
AssertionError: get_apps_modules should only return normal Applications
----------------------------------------------------------------------
```

--- a/corehq/apps/app_manager/models.py
+++ b/corehq/apps/app_manager/models.py
@@ -3573,7 +3573,7 @@ class AdvancedModule(ModuleBase):
                 )
 
                 if from_module.parent_select.active:
-                    from_app = from_module.get_app()
+                    from_app = from_module.get_app()  # A form can be copied from a module in a different app.
                     select_chain = get_select_chain(from_app, from_module, include_self=False)
                     for n, link in enumerate(reversed(list(enumerate(select_chain)))):
                         i, module = link

--- a/corehq/apps/app_manager/static/app_manager/js/forms/copy_form_to_app.js
+++ b/corehq/apps/app_manager/static/app_manager/js/forms/copy_form_to_app.js
@@ -19,7 +19,7 @@ hqDefine("app_manager/js/forms/copy_form_to_app", function () {
             })
         );
         return self;
-    }
+    };
 
     var appsModulesModel = function (appsModules) {
         var self = {};
@@ -30,10 +30,9 @@ hqDefine("app_manager/js/forms/copy_form_to_app", function () {
         );
         self.selectedAppId = ko.observable(self.apps()[0].id);
         self.selectedApp = ko.computed(function () {
-            var selected = _.filter(self.apps(), function (app) {
+            return _.find(self.apps(), function (app) {
                 return app.id === self.selectedAppId();
             });
-            return selected.length ? selected[0] : undefined;
         });
         return self;
     };

--- a/corehq/apps/app_manager/static/app_manager/js/forms/copy_form_to_app.js
+++ b/corehq/apps/app_manager/static/app_manager/js/forms/copy_form_to_app.js
@@ -2,22 +2,28 @@
 hqDefine("app_manager/js/forms/copy_form_to_app", function () {
     "use strict";
 
-    var module = function (moduleId, moduleName) {
+    var module = function (moduleId, moduleName, isCurrentModule) {
         var self = {};
         self.id = moduleId;
         self.name = moduleName;
+        self.isCurrent = isCurrentModule;
         return self;
     };
 
-    var application = function (appId, appName, modules) {
+    var application = function (appId, appName, isCurrentApp, modules) {
         var self = {};
         self.id = appId;
         self.name = appName;
+        self.isCurrent = isCurrentApp;
         self.modules = ko.observableArray(
             _.map(modules, function (mod) {
-                return module(mod["module_id"], mod["name"]);
+                return module(mod["module_id"], mod["name"], mod["is_current"]);
             })
         );
+        var currentModule = _.find(self.modules(), function (mod) {
+            return mod.isCurrent;
+        });
+        self.selectedModuleId = ko.observable(currentModule ? currentModule.id : undefined);
         return self;
     };
 
@@ -25,10 +31,13 @@ hqDefine("app_manager/js/forms/copy_form_to_app", function () {
         var self = {};
         self.apps = ko.observableArray(
             _.map(appsModules, function (app) {
-                return application(app["app_id"], app["name"], app["modules"]);
+                return application(app["app_id"], app["name"], app["is_current"], app["modules"]);
             })
         );
-        self.selectedAppId = ko.observable(self.apps()[0].id);
+        var currentApp = _.find(self.apps(), function (app) {
+            return app.isCurrent;
+        });
+        self.selectedAppId = ko.observable(currentApp ? currentApp.id : undefined);
         self.selectedApp = ko.computed(function () {
             return _.find(self.apps(), function (app) {
                 return app.id === self.selectedAppId();

--- a/corehq/apps/app_manager/templates/app_manager/partials/form_tab_advanced.html
+++ b/corehq/apps/app_manager/templates/app_manager/partials/form_tab_advanced.html
@@ -27,7 +27,8 @@
                                 style="min-width: 150px;"
                                 data-bind="options: modules,
                                            optionsText: 'name',
-                                           optionsValue: 'id'"></select>
+                                           optionsValue: 'id',
+                                           value: selectedModuleId"></select>
                         <!-- /ko -->
                     </span>
                     {% else %}

--- a/corehq/apps/app_manager/tests/test_suite.py
+++ b/corehq/apps/app_manager/tests/test_suite.py
@@ -255,14 +255,10 @@ class SuiteTest(SimpleTestCase, TestXmlMixin, SuiteMixin):
 
         src_app.copy_form(src_module, original_form, dst_module, rename=True)
 
-        form_count = 0
-        for f in src_app.get_forms():
-            form_count += 1
-        self.assertEqual(form_count, 1, 'Form copied to the wrong app')
-        form_count = 0
-        for f in dst_app.get_forms():
-            form_count += 1
-        self.assertEqual(form_count, 1)
+        self.assertEqual(len(list(src_app.get_forms())), 1, 'Form copied to the wrong app')
+        dst_app_forms = list(dst_app.get_forms())
+        self.assertEqual(len(dst_app_forms), 1)
+        self.assertEqual(dst_app_forms[0].name['en'], 'Copy of Untitled Form')
 
     def test_owner_name(self):
         self._test_generic_suite('owner-name')

--- a/corehq/apps/app_manager/tests/test_views.py
+++ b/corehq/apps/app_manager/tests/test_views.py
@@ -237,6 +237,7 @@ class TestViews(TestCase):
         deleted_app.add_module(Module.new_module("Module0", "en"))
         deleted_app.save()
         deleted_app.delete_app()
+        deleted_app.save()
         self.addCleanup(lambda: Application.get_db().delete_doc(deleted_app.id))
 
         apps_modules = get_apps_modules(self.domain.name)
@@ -247,6 +248,6 @@ class TestViews(TestCase):
             'Each app should only have one module'
         )
         self.assertEqual(
-            apps_modules[0]['modules']['name'], {'en': 'Module0'},
-            'Module name should be a dictionary'
+            apps_modules[0]['modules'][0]['name'], 'Module0',
+            'Module name should be translated'
         )

--- a/corehq/apps/app_manager/tests/test_views.py
+++ b/corehq/apps/app_manager/tests/test_views.py
@@ -36,18 +36,18 @@ class TestViews(TestCase):
     @classmethod
     def setUpClass(cls):
         super(TestViews, cls).setUpClass()
-        cls.domain = Domain(name='app-manager-testviews-domain', is_active=True)
-        cls.domain.save()
+        cls.project = Domain(name='app-manager-testviews-domain', is_active=True)
+        cls.project.save()
         cls.username = 'cornelius'
         cls.password = 'fudge'
-        cls.user = WebUser.create(cls.domain.name, cls.username, cls.password, is_active=True)
+        cls.user = WebUser.create(cls.project.name, cls.username, cls.password, is_active=True)
         cls.user.is_superuser = True
         cls.user.save()
         cls.build = add_build(version='2.7.0', build_number=20655)
-        toggles.CUSTOM_PROPERTIES.set("domain:{domain}".format(domain=cls.domain.name), True)
+        toggles.CUSTOM_PROPERTIES.set("domain:{domain}".format(domain=cls.project.name), True)
 
     def setUp(self):
-        self.app = Application.new_app(self.domain.name, "TestApp")
+        self.app = Application.new_app(self.project.name, "TestApp")
         self.app.build_spec = BuildSpec.from_string('2.7.0/latest')
         self.client.login(username=self.username, password=self.password)
 
@@ -82,13 +82,13 @@ class TestViews(TestCase):
         self.app.save()
 
         mock.side_effect = XFormValidationError('')
-        response = self.client.get(reverse('app_download_file', kwargs=dict(domain=self.domain.name,
+        response = self.client.get(reverse('app_download_file', kwargs=dict(domain=self.project.name,
                                                                             app_id=self.app.get_id,
                                                                             path='modules-0/forms-0.xml')))
         self.assertEqual(response.status_code, 404)
 
     def test_edit_commcare_profile(self, mock):
-        app = Application.new_app(self.domain.name, "TestApp")
+        app = Application.new_app(self.project.name, "TestApp")
         app.save()
         data = {
             "custom_properties": {
@@ -97,7 +97,7 @@ class TestViews(TestCase):
             }
         }
 
-        response = self.client.post(reverse('edit_commcare_profile', args=[self.domain.name, app._id]),
+        response = self.client.post(reverse('edit_commcare_profile', args=[self.project.name, app._id]),
                                     json.dumps(data),
                                     content_type='application/json')
 
@@ -113,7 +113,7 @@ class TestViews(TestCase):
             }
         }
 
-        response = self.client.post(reverse('edit_commcare_profile', args=[self.domain.name, app._id]),
+        response = self.client.post(reverse('edit_commcare_profile', args=[self.project.name, app._id]),
                                     json.dumps(data),
                                     content_type='application/json')
 
@@ -138,7 +138,7 @@ class TestViews(TestCase):
         self.app.save()
 
         kwargs = {
-            'domain': self.domain.name,
+            'domain': self.project.name,
             'app_id': self.app.id,
         }
         self._test_status_codes([
@@ -150,19 +150,19 @@ class TestViews(TestCase):
         build = self.app.make_build()
         build.save()
         content = self._json_content_from_get('current_app_version', {
-            'domain': self.domain.name,
+            'domain': self.project.name,
             'app_id': self.app.id,
         })
         self.assertEqual(content['currentVersion'], 1)
         self.app.save()
         content = self._json_content_from_get('current_app_version', {
-            'domain': self.domain.name,
+            'domain': self.project.name,
             'app_id': self.app.id,
         })
         self.assertEqual(content['currentVersion'], 2)
 
         content = self._json_content_from_get('paginate_releases', {
-            'domain': self.domain.name,
+            'domain': self.project.name,
             'app_id': self.app.id,
         }, {'limit': 5})
         self.assertEqual(len(content), 1)
@@ -180,7 +180,7 @@ class TestViews(TestCase):
         module = self.app.add_module(AdvancedModule.new_module("Module0", "en"))
         self.app.save()
         self._test_status_codes(['view_module'], {
-            'domain': self.domain.name,
+            'domain': self.project.name,
             'app_id': self.app.id,
             'module_unique_id': module.unique_id,
         })
@@ -189,7 +189,7 @@ class TestViews(TestCase):
         module = self.app.add_module(ReportModule.new_module("Module0", "en"))
         self.app.save()
         self._test_status_codes(['view_module'], {
-            'domain': self.domain.name,
+            'domain': self.project.name,
             'app_id': self.app.id,
             'module_unique_id': module.unique_id,
         })
@@ -198,7 +198,7 @@ class TestViews(TestCase):
         module = self.app.add_module(ShadowModule.new_module("Module0", "en"))
         self.app.save()
         self._test_status_codes(['view_module'], {
-            'domain': self.domain.name,
+            'domain': self.project.name,
             'app_id': self.app.id,
             'module_unique_id': module.unique_id,
         })
@@ -206,12 +206,12 @@ class TestViews(TestCase):
     def test_dashboard(self, mock):
         # This redirects to the dashboard
         self._test_status_codes(['default_app'], {
-            'domain': self.domain.name,
+            'domain': self.project.name,
         }, True)
 
     def test_default_new_app(self, mock):
         response = self.client.get(reverse('default_new_app', kwargs={
-            'domain': self.domain.name,
+            'domain': self.project.name,
         }), follow=False)
 
         self.assertEqual(response.status_code, 302)
@@ -225,22 +225,22 @@ class TestViews(TestCase):
         self.app.add_module(Module.new_module("Module0", "en"))
         self.app.save()
 
-        other_app = Application.new_app(self.domain.name, "OtherApp")
+        other_app = Application.new_app(self.project.name, "OtherApp")
         other_app.add_module(Module.new_module("Module0", "en"))
         other_app.save()
         self.addCleanup(lambda: Application.get_db().delete_doc(other_app.id))
 
-        linked_app = create_linked_app(self.domain.name, self.app.id, self.domain.name, 'LinkedApp')
+        linked_app = create_linked_app(self.project.name, self.app.id, self.project.name, 'LinkedApp')
         self.addCleanup(lambda: Application.get_db().delete_doc(linked_app.id))
 
-        deleted_app = Application.new_app(self.domain.name, "DeletedApp")
+        deleted_app = Application.new_app(self.project.name, "DeletedApp")
         deleted_app.add_module(Module.new_module("Module0", "en"))
         deleted_app.save()
         deleted_app.delete_app()
         deleted_app.save()
         self.addCleanup(lambda: Application.get_db().delete_doc(deleted_app.id))
 
-        apps_modules = get_apps_modules(self.domain.name)
+        apps_modules = get_apps_modules(self.project.name)
 
         self.assertEqual(len(apps_modules), 2, 'get_apps_modules should only return normal Applications')
         self.assertTrue(

--- a/corehq/apps/app_manager/views/forms.py
+++ b/corehq/apps/app_manager/views/forms.py
@@ -672,10 +672,7 @@ def get_form_view_context_and_template(request, domain, form, langs, messages=me
     if toggles.CUSTOM_ICON_BADGES.enabled(domain):
         context['form_icon'] = form.custom_icon if form.custom_icon else CustomIcon()
 
-    if (
-            toggles.COPY_FORM_TO_APP.enabled(domain) or
-            toggles.COPY_FORM_TO_APP.enabled(request.user.username)
-    ):
+    if toggles.COPY_FORM_TO_APP.enabled_for_request(request):
         context['apps_modules'] = get_apps_modules(domain)
 
     if context['allow_form_workflow'] and toggles.FORM_LINK_WORKFLOW.enabled(domain):

--- a/corehq/apps/app_manager/views/forms.py
+++ b/corehq/apps/app_manager/views/forms.py
@@ -511,17 +511,22 @@ def get_form_questions(request, domain, app_id):
     return json_response(xform_questions)
 
 
-def get_apps_modules(domain):
+def get_apps_modules(domain, current_app_id=None, current_module_id=None):
     """
-    Returns a domain's apps and their modules
+    Returns a domain's apps and their modules.
+
+    If current_app_id and current_module_id are given, "is_current" is
+    set to True for them.
     """
     return [
         {
             'app_id': app.id,
             'name': app.name,
+            'is_current': app.id == current_app_id,
             'modules': [{
                 'module_id': module.id,
                 'name': clean_trans(module.name, app.langs),
+                'is_current': module.unique_id == current_module_id,
             } for module in app.modules]
         }
         for app in get_apps_in_domain(domain, include_remote=False)
@@ -673,7 +678,7 @@ def get_form_view_context_and_template(request, domain, form, langs, messages=me
         context['form_icon'] = form.custom_icon if form.custom_icon else CustomIcon()
 
     if toggles.COPY_FORM_TO_APP.enabled_for_request(request):
-        context['apps_modules'] = get_apps_modules(domain)
+        context['apps_modules'] = get_apps_modules(domain, app.id, module.unique_id)
 
     if context['allow_form_workflow'] and toggles.FORM_LINK_WORKFLOW.enabled(domain):
         def qualified_form_name(form, auto_link):

